pavelib/tests.py

```diff
@@ -66,28 +70,61 @@ def test_system(options, passthrough_options):
     """
     system = getattr(options, 'system', None)
     test_id = getattr(options, 'test_id', None)
+    django_version = getattr(options, 'django_version', None)
+
+    assert(system in (None, 'lms', 'cms'))
+    assert(django_version in (None, '1.8', '1.11'))
+
+    def _create_tox_system_test_suites(systems):
+        system_tests = []
+        test_paths = []
+        # Add all test directories for the specified systems.
+        for syst in systems:
+            test_paths.extend(suites.default_system_test_dirs(syst))
+        # Remove all duplicate paths.
+        test_paths = list(set(test_paths))
+        for test_id in test_paths:
+            system = test_id.split('/')[0]
+            syst = 'cms'
+            if system in ['common', 'openedx', 'lms']:
+                syst = 'lms'
+            system_tests.append(suites.SystemTestSuite(
+                syst,
+                passthrough_options=passthrough_options,
+                test_id=test_id,
+                **options.test_system
+            ))
+        return system_tests
 
     if test_id:
+        # Testing a single test ID.
+        # Ensure the proper system for the test id.
         if not system:
             system = test_id.split('/')[0]
         if system in ['common', 'openedx']:
             system = 'lms'
-        options.test_system['test_id'] = test_id
-
-    if test_id or system:
         system_tests = [suites.SystemTestSuite(
             system,
             passthrough_options=passthrough_options,
             **options.test_system
         )]
     else:
-        system_tests = []
-        for syst in ('cms', 'lms'):
-            system_tests.append(suites.SystemTestSuite(
-                syst,
-                passthrough_options=passthrough_options,
-                **options.test_system
-            ))
+        # Testing a single system -or- both systems.
+        if system:
+            systems = [system]
+        else:
+            # No specified system or test_id, so test all tests of both systems.
+            systems = ['cms', 'lms']
+        if django_version:
+            system_tests = _create_tox_system_test_suites(systems)
```

Comment:
So, to clarify, this means we run tox against one test_id at a time, vs. bunching them up in one command?

Reply:
The code creates a separate subsuite for each top-level test directory. Each subsuite is executed in a separate pytest command in the newly built tox venv.

Reply:
With the way pytest test discovery and its Python root determination works, a statement like:
`pytest common/lib lms/djangoapps`
does not currently work due the import paths assuming that several Python roots exist (as they do when the application is run within Django). Only single top-level test directories have worked when I've tried it. @jmbowman Correct me if I'm wrong.

Reply:
I think pytest should work with the same groupings we've supported in paver to date (that's how I decided which sets of tests should have separate settings).  You can't run them all in a single pytest run and expect them to pass, but paver didn't allow that anyway.  And I think the current setup is explicitly designed to run the common/openedx tests with both lms and cms settings to make sure they work for both services.

Reply:
@jmbowman Do you have a list of those groupings? Currently, when running `paver test_system --django_version=1.11`, the following tox commands are executed:
```
tox -e py27-django111 -- --ds=lms.envs.test --junitxml=/edx/app/edxapp/edx-platform/reports/lms/nosetests.xml -p no:randomly openedx/core/lib --cov --cov-report=
tox -e py27-django111 -- --ds=lms.envs.test --junitxml=/edx/app/edxapp/edx-platform/reports/lms/nosetests.xml -p no:randomly openedx/core/djangolib --cov --cov-report=
tox -e py27-django111 -- --ds=lms.envs.test --junitxml=/edx/app/edxapp/edx-platform/reports/lms/nosetests.xml -p no:randomly openedx/features --cov --cov-report=
tox -e py27-django111 -- --ds=lms.envs.test --junitxml=/edx/app/edxapp/edx-platform/reports/lms/nosetests.xml -p no:randomly lms/djangoapps --cov --cov-report=
tox -e py27-django111 -- --ds=lms.envs.test --junitxml=/edx/app/edxapp/edx-platform/reports/lms/nosetests.xml -p no:randomly openedx/core/djangoapps --cov --cov-report=
tox -e py27-django111 -- --ds=lms.envs.test --junitxml=/edx/app/edxapp/edx-platform/reports/lms/nosetests.xml -p no:randomly openedx/tests --cov --cov-report=
tox -e py27-django111 -- --ds=cms.envs.test --junitxml=/edx/app/edxapp/edx-platform/reports/cms/nosetests.xml -p no:randomly cms/lib --cov --cov-report=
tox -e py27-django111 -- --ds=lms.envs.test --junitxml=/edx/app/edxapp/edx-platform/reports/lms/nosetests.xml -p no:randomly lms/lib --cov --cov-report=
tox -e py27-django111 -- --ds=cms.envs.test --junitxml=/edx/app/edxapp/edx-platform/reports/cms/nosetests.xml -p no:randomly cms/djangoapps --cov --cov-report=
tox -e py27-django111 -- --ds=lms.envs.test --junitxml=/edx/app/edxapp/edx-platform/reports/lms/nosetests.xml -p no:randomly lms/tests.py --cov --cov-report=
tox -e py27-django111 -- --ds=lms.envs.test --junitxml=/edx/app/edxapp/edx-platform/reports/lms/nosetests.xml -p no:randomly common/djangoapps --cov --cov-report=
```
Is the above sufficient/optimal?

Reply:
@pwnage101 After some testing, seems it's fine to move back to the test collection method as the non-tox code path. So I've done so. Re-review?